TESTS/mbed_platform/FileHandle/main.cpp

```diff
@@ -0,0 +1,532 @@
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_EQUAL_INT(str1_size, read_ret);
+    TEST_ASSERT_EQUAL_INT(0, strncmp(str1, read_buf, str1_size));
+
+#ifndef __ICCARM__
+    // read 3; expected read 2
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_TRUE(std::feof(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(str2_size - 1, read_ret);
+    TEST_ASSERT_EQUAL_INT(0, strncmp(str2, read_buf, str2_size - 1));
+
+    // read 3; expected read 0
+    TestFile<FS>::resetFunctionCallHistory();
+    read_ret = std::fread(read_buf, 1, str2_size, file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnRead));
+    TEST_ASSERT_TRUE(std::feof(file) > 0);
+    std::clearerr(file); // for ARMCC
+    TEST_ASSERT_EQUAL_INT(0, read_ret);
+#endif
+
+    std::fclose(file);
+}
+
+/** Test fputc and fgetc
+ *
+ *  Given already opened file
+ *
+ *  When write some data to file
+ *  Then underneath retargeting layer write function is called
+ *       fputc return written element
+ *       on failure, returns EOF and stream error is sets
+ *
+ *  When read previously written data from file
+ *  Then underneath retargeting layer read function is called
+ *       fgetc return read element
+ *       read data match previously written
+ *       on failure, returns EOF and stream error/eof is sets respectively
+ *
+ */
+void test_fputc_fgetc()
+{
+    std::FILE *file;
+    const uint32_t FS = 3;
+    TestFile<FS> fh;
+    char char_buf[3] = {'a', 'b', 'c' };
+    int ret;
+
+    file = fdopen(&fh, "w+");
+    TEST_ASSERT_NOT_NULL(file);
+#ifdef __ICCARM__
+    std::setbuf(file, NULL); // fix IAR every byte sync
+#else
+    mbed_set_unbuffered_stream(file);
+#endif
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[0], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[0], ret);
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[1], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[1], ret);
+
+    // write 1; expected written 1
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[2], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_EQUAL_INT(char_buf[2], ret);
+
+#ifndef __ICCARM__ // prevents IAR infinite loop
+    // write 1; expected written 0
+    TestFile<FS>::resetFunctionCallHistory();
+    ret = std::fputc(char_buf[0], file);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnWrite));
+    TEST_ASSERT_TRUE(std::ferror(file) > 0);
+    std::clearerr(file); // for ARMCC
```

Comment:
You keep calling `fseek(file, 0, SEEK_SET)`. If you called `rewind()` instead, that would clear the error. Should be necessary on every compiler

Reply:
fixed